crates/ide/src/inlay_hints/adjustment.rs

```diff
@@ -113,19 +146,117 @@ pub(super) fn hints(
     Some(())
 }
 
+/// Returns whatever the hint should be postfix and if we need to add paretheses on the inside and/or outside of `expr`,
+/// if we are going to add (`postfix`) adjustments hints to it.
+fn mode_and_needs_parens_for_adjustment_hints(
+    expr: &ast::Expr,
+    mode: AdjustmentHintsMode,
+) -> (bool, bool, bool) {
+    use {std::cmp::Ordering::*, AdjustmentHintsMode::*};
+
+    match mode {
+        Prefix | Postfix => {
+            let postfix = matches!(mode, Postfix);
+            let (inside, outside) = needs_parens_for_adjustment_hints(expr, postfix);
+            (postfix, inside, outside)
+        }
+        PreferPrefix | PreferPostfix => {
+            let prefer_postfix = matches!(mode, PreferPostfix);
+
+            let (pre_inside, pre_outside) = needs_parens_for_adjustment_hints(expr, false);
+            let prefix = (false, pre_inside, pre_outside);
+            let pre_count = pre_inside as u8 + pre_outside as u8;
+
+            let (post_inside, post_outside) = needs_parens_for_adjustment_hints(expr, true);
+            let postfix = (true, post_inside, post_outside);
+            let post_count = post_inside as u8 + post_outside as u8;
+
+            match pre_count.cmp(&post_count) {
+                Less => prefix,
+                Greater => postfix,
+                Equal if prefer_postfix => postfix,
+                Equal => prefix,
+            }
+        }
+    }
+}
+
+/// Returns whatever we need to add paretheses on the inside and/or outside of `expr`,
+/// if we are going to add (`postfix`) adjustments hints to it.
+fn needs_parens_for_adjustment_hints(expr: &ast::Expr, postfix: bool) -> (bool, bool) {
+    // This is a very miserable pile of hacks...
+    //
+    // `Expr::needs_parens_in` requires that the expression is the child of the other expression,
+    // that is supposed to be its parent.
+    //
+    // But we want to check what would happen if we add `*`/`.*` to the inner expression.
+    // To check for inner we need `` expr.needs_parens_in(`*expr`) ``,
+    // to check for outer we need `` `*expr`.needs_parens_in(parent) ``,
+    // where "expr" is the `expr` parameter, `*expr` is the editted `expr`,
+    // and "parent" is the parent of the original expression...
+    //
+    // For this we utilize mutable mutable trees, which is a HACK, but it works.
```

Comment:
Ye I am not fond of this at all to be honest :) But I also don't really want to block the PR on this since I can't think of a better way right now either ...
Can you put a FIXME on this/on `needs_parens_in` API that this is not ideal? Or maybe open an issue about that, either works.